sdks/java/core/src/test/java/org/apache/beam/sdk/runners/PipelineRunnerTest.java

```diff
@@ -39,4 +49,36 @@ public void testInstantiation() {
     PipelineRunner<?> runner = PipelineRunner.fromOptions(options);
     assertTrue(runner instanceof CrashingRunner);
   }
+
+  private static class ScaleFn<T extends Number>
+      extends SimpleFunction<T, Double> {
+
+    private final double scalar;
+
+    public ScaleFn(double scalar) {
+      this.scalar = scalar;
+    }
+
+    @Override
+    public Double apply(T input) {
+      return scalar * input.doubleValue();
+    }
+  }
+
+  @Test
+  @Category(ValidatesRunner.class)
+  public void testRunPTransform() {
+    PipelineRunner pipelineRunner = PipelineRunner.create();
+    PTransform<PBegin, POutput> scaleByTwo = new PTransform<PBegin, POutput>() {
+      @Override
+      public POutput expand(PBegin input) {
+        PCollection<Double> output = input
+            .apply(Create.<Integer>of(1, 2, 3, 4))
+            .apply("ScaleByTwo", MapElements.via(new ScaleFn<Integer>(2.0)));
+        PAssert.that(output).containsInAnyOrder(2.0, 4.0, 6.0, 8.0);
+        return output;
+      }
+    };
+    pipelineRunner.run(scaleByTwo);
+  }
```

Comment:
There is something slightly unsatisfying about this test in that we don't know if the pipeline actually ran... Maybe you could have your ScaleFn increment a single counter for every input it sees, which you would check for out of the result, similar to (but greatly simplified from)

https://github.com/apache/beam/blob/61777b4338733d99f4f858be0d7d0313ec138a06/sdks/java/core/src/test/java/org/apache/beam/sdk/metrics/MetricsTest.java#L233

Reply:
Done. So PAssert.that() doesn't really determine whether the pipeline ran?

Reply:
Correct. A the end of the day it's constructs a DoFn that fails if the assert fails (which, of course, is more motivation for the very change you're working on, 'cause PAsserts don't do anything if run() is not called).